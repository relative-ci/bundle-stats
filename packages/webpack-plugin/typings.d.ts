import { ReportOptions } from '@bundle-stats/cli-utils';

export class BundleStatsWebpackPlugin {
  constructor(options?: Partial<BundleStatsWebpackPlugin.Options>) {}

  apply(compiler: BundleStatsWebpackPlugin.Compiler): void;
}

declare namespace BundleStatsWebpackPlugin {
  interface WebpackStatsOptions {
    /**
     * Output webpack assets information
     * Default: `true`.
     */
    assets?: Boolean;

    /**
     * Output webpack chunks information
     * Default: `true`.
     */
    chunks?: Boolean;

    /**
     * Output webpack modules information
     * Default: `true`.
     */
    modules?: Boolean;

    /**
     * Output webpack hash information
     * Default: `true`.
     */
    hash?: Boolean;

    /**
     * Output webpack builtAt information
     * Default: `true`.
     */
    builtAt?: Boolean;
  }

  interface Options extends ReportOptions {
    /**
     * webpack stats options
     * Default: `{ assets: true: true, chunks: true, modules: true, hash: true, builtAt: true }`.
     */
    stats?: Partial<WebpackStatsOptions>;
  }

  type Compiler = import('webpack').Compiler;
}
