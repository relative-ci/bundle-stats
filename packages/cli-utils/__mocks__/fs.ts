const { vol } = require('memfs');

module.exports = vol;
