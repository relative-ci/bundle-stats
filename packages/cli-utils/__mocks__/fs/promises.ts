const { vol } = require('memfs');

module.exports = vol.promises;
