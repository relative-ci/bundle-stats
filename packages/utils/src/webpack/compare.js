import { MetricTypes } from '../constants';
import { compareMetrics } from '../report/compare-metrics';
import {
  SECTION_WEBPACK_STATS,
  SECTION_WEBPACK_SIZES,
  SECTION_WEBPACK_ASSETS,
  SECTION_WEBPACK_MODULES,
  SECTION_WEBPACK_PACKAGES,
  SECTIONS,
} from './constants';
import { selectors } from './selectors';

/**
 * Compare stats metrics
 *
 * @param {Object[]} jobs - List of jobs to compare
 * @param {Array<Function>} [rowTransformers]
 *
 * @return {Object[]} Compared stats metrics
 */
const compareStats = (jobs, rowTransformers) =>
  compareMetrics(jobs, selectors.stats, undefined, rowTransformers);

/**
 * Compare size metrics
 *
 * @param {Object[]} jobs - List of jobs to compare
 * @param {Array<Function>} [rowTransformers]
 *
 * @return {Object[]} Compared size metrics
 */
const compareSizes = (jobs, rowTransformers) =>
  compareMetrics(jobs, selectors.sizes, undefined, rowTransformers);

/**
 * Compare asset metrics
 *
 * @param {Object[]} jobs - List of jobs to compare
 * @param {Array<Function>} [rowTransformers]
 *
 * @return {Object[]} Compared asset metrics
 */
const compareAssets = (jobs, rowTransformers) =>
  compareMetrics(jobs, selectors.assets, MetricTypes.FileSize, rowTransformers);

/**
 * Compare all modules
 *
 * @param {Object[]} jobs - List of jobs to compare
 * @param {Array<Function>} [rowTransformers]
 * @return {Object[]} Compared module metrics
 */
const compareModules = (jobs, rowTransformers) =>
  compareMetrics(jobs, selectors.modules, MetricTypes.FileSize, rowTransformers);

/**
 * Compare package metrics
 *
 * @param {Object[]} jobs - List of jobs to compare
 * @param {Array<Function>} [rowTransformers]
 *
 * @return {Object[]} Compared package metrics
 */
const comparePackages = (jobs, rowTransformers) =>
  compareMetrics(jobs, selectors.packages, MetricTypes.FileSize, rowTransformers);

export const compareBySection = {
  [SECTION_WEBPACK_STATS]: compareStats,
  [SECTION_WEBPACK_SIZES]: compareSizes,
  [SECTION_WEBPACK_ASSETS]: compareAssets,
  [SECTION_WEBPACK_MODULES]: compareModules,
  [SECTION_WEBPACK_PACKAGES]: comparePackages,
};

/**
 * Compare webpack sections
 *
 * @param {Object[]} jobs - List of jobs to compare
 * @return {Object} Compared metrics by section
 */
export const compare = (jobs) =>
  SECTIONS.reduce(
  (agg, sectionId) => ({
    ...agg,
    [sectionId]: compareBySection[sectionId](jobs),
  }),
  {},
);
