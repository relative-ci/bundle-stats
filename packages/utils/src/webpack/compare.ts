import { MetricTypes } from '../constants';
import { compareMetrics } from '../report/compare-metrics';
import { MetricReportRowTransformFn } from '../report/types';
import {
  SECTION_WEBPACK_STATS,
  SECTION_WEBPACK_SIZES,
  SECTION_WEBPACK_ASSETS,
  SECTION_WEBPACK_MODULES,
  SECTION_WEBPACK_PACKAGES,
  SECTIONS,
} from './constants';
import { selectors, getModulesDuplicateSizeMetrics, getModulesTotalSizeMetrics } from './selectors';

const compareStats = (jobs: Array<unknown>, rowTransformers?: Array<MetricReportRowTransformFn>) =>
  compareMetrics(jobs, selectors.stats, undefined, rowTransformers);

const compareSizes = (jobs: Array<unknown>, rowTransformers?: Array<MetricReportRowTransformFn>) =>
  compareMetrics(jobs, selectors.sizes, undefined, rowTransformers);

const compareAssets = (jobs: Array<unknown>, rowTransformers?: Array<MetricReportRowTransformFn>) =>
  compareMetrics(jobs, selectors.assets, MetricTypes.FileSize, rowTransformers);

const compareModules = (
  jobs: Array<unknown>,
  rowTransformers?: Array<MetricReportRowTransformFn>,
) => compareMetrics(jobs, selectors.modules, MetricTypes.FileSize, rowTransformers);

export const compareModuleDuplicateSize = (
  jobs: Array<unknown>,
  rowTransformers?: Array<MetricReportRowTransformFn>,
) => compareMetrics(jobs, getModulesDuplicateSizeMetrics, MetricTypes.FileSize, rowTransformers);

export const compareModuleTotalSize = (
  jobs: Array<unknown>,
  rowTransformers?: Array<MetricReportRowTransformFn>,
) => compareMetrics(jobs, getModulesTotalSizeMetrics, MetricTypes.FileSize, rowTransformers);

const comparePackages = (
  jobs: Array<unknown>,
  rowTransformers?: Array<MetricReportRowTransformFn>,
) => compareMetrics(jobs, selectors.packages, MetricTypes.FileSize, rowTransformers);

export const compareBySection = {
  [SECTION_WEBPACK_STATS]: compareStats,
  [SECTION_WEBPACK_SIZES]: compareSizes,
  [SECTION_WEBPACK_ASSETS]: compareAssets,
  [SECTION_WEBPACK_MODULES]: compareModules,
  [SECTION_WEBPACK_PACKAGES]: comparePackages,
};

/* eslint-disable prettier/prettier */
/**
 * Compare webpack sections
 */
export const compare = (jobs: Array<unknown>) => SECTIONS.reduce(
  (agg, sectionId) => ({
    ...agg,
    [sectionId]: compareBySection[sectionId as keyof typeof compareBySection](jobs),
  }),
  {},
);
/* eslint-enable prettier/prettier */
