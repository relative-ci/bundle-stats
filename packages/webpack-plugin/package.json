{
  "name": "bundle-stats-webpack-plugin",
  "version": "4.5.0",
  "description": "In-depth bundle analyzer for webpack(bundle size, assets, modules, packages)",
  "main": "lib/index.js",
  "types": "typings.d.ts",
  "directories": {
    "lib": "lib"
  },
  "engines": {
    "node": ">= 14.0"
  },
  "scripts": {
    "build": "babel src -d lib",
    "lint": "../../node_modules/.bin/eslint . --ext .js",
    "test": "TZ=UTC jest src ",
    "pretest:package": "cd test/package && npm install --force",
    "test:package": "cd test/package && npm test --"
  },
  "keywords": [
    "webpack",
    "webpack-plugin",
    "bundle-size",
    "bundle-analyzer",
    "bundle-stats",
    "stats",
    "bundle",
    "size",
    "assets",
    "chunks",
    "modules"
  ],
  "author": {
    "name": "Viorel Cojocaru",
    "email": "vio@beanon.com",
    "url": "https://beanon.com"
  },
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/relative-ci/bundle-stats.git"
  },
  "bugs": {
    "url": "https://github.com/relative-ci/bundle-stats/issues"
  },
  "homepage": "https://github.com/relative-ci/bundle-stats/blob/master/packages/webpack-plugin#readme",
  "dependencies": {
    "@bundle-stats/cli-utils": "^4.5.0",
    "core-js": "^3.21.0",
    "lodash": "^4.17.0"
  },
  "devDependencies": {
    "@babel/cli": "7.22.9",
    "@babel/core": "7.22.9",
    "@babel/preset-env": "7.22.9",
    "@types/jest": "29.5.3",
    "jest": "29.6.1",
    "jest-date-mock": "1.0.8",
    "memory-fs": "0.5.0",
    "webpack": "5.88.2"
  },
  "peerDependencies": {
    "webpack": "^4.0.0 || ^5.0.0"
  }
}
