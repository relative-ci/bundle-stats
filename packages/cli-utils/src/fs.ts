import { createReadStream } from 'fs';
import { writeFile, stat } from 'fs/promises';
import { parser } from 'stream-json';
import { chain } from 'stream-chain';
import Asm from 'stream-json/Assembler';

export const readJSONStream = async <T = unknown>(filepath: string): Promise<T> => {
  // Check if the file exists and throw error before creating a stream
  await stat(filepath);

  const readStream = createReadStream(filepath);
  const pipeline = chain([readStream, parser()]);
  const asm = Asm.connectTo(pipeline);

  return new Promise((resolve, reject) => {
    asm.on('done', (data) => {
      if (data.current) {
        resolve(data.current);
      } else {
        reject(new Error('Invalid JSON file'));
      }
    });
  });
};

export async function writeJSON(filepath: string, data: Record<string, unknown>): Promise<void> {
  return writeFile(filepath, JSON.stringify(data));
}
