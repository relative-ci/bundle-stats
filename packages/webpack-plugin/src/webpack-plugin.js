import webpack from 'webpack';
import { merge } from 'lodash';
import { generateReports } from '@bundle-stats/cli-utils';

import * as CONFIG from './config';

const DEFAULT_OPTIONS = {
  baseline: Boolean(process.env.BUNDLE_STATS_BASELINE),
  stats: {
    assets: true,
    chunks: true,
    modules: true,
    hash: true,
    builtAt: true,
  },
};

const PLUGIN_NAME = 'BundleStats';

const isWebpack5 = parseInt(webpack.version, 10) === 5;

export class BundleStatsWebpackPlugin {
  constructor(options) {
    this.options = options;
  }

  apply(compiler) {
    const options = merge({}, DEFAULT_OPTIONS, this.options);

    if (isWebpack5) {
      compiler.hooks.thisCompilation.tap(PLUGIN_NAME, (compilation) => {
        compilation.hooks.processAssets.tapPromise(
          {
            name: PLUGIN_NAME,
            stage: webpack.Compilation.PROCESS_ASSETS_STAGE_REPORT,
          },
          async () => {
            const newAssets = await generateReports(
              compilation.getStats().toJson(options.stats),
              options,
              {
                outputPath: compilation?.options?.output?.path,
                invalidOptionsUrl: CONFIG.OPTIONS_URL,
                logger:
                  compilation.getInfrastructureLogger &&
                  compilation.getInfrastructureLogger(PLUGIN_NAME),
              },
            );

            Object.entries(newAssets).forEach(([filename, source]) => {
              compilation.emitAsset(filename, new webpack.sources.RawSource(source), {
                development: true,
              });
            });
          },
        );
      });

      return;
    }

    compiler.hooks.emit.tapAsync(PLUGIN_NAME, async (compilation, callback) => {
      const newAssets = await generateReports(
        compilation.getStats().toJson(options.stats),
        options,
        {
          outputPath: compilation?.options?.output?.path,
          invalidOptionsUrl: CONFIG.OPTIONS_URL,
          logger:
            compilation.getInfrastructureLogger && compilation.getInfrastructureLogger(PLUGIN_NAME),
        },
      );

      Object.entries(newAssets).forEach(([filename, source]) => {
        // eslint-disable-next-line no-param-reassign
        compilation.assets[filename] = {
          size: () => 0,
          source: () => source,
        };
      });

      callback();
    });
  }
}
