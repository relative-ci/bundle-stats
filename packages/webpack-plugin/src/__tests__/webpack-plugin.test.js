import path from 'path';
import fs from 'fs/promises';
import webpack from 'webpack';
import MemoryFS from 'memory-fs';
import { advanceTo } from 'jest-date-mock';

import { BundleStatsWebpackPlugin } from '../webpack-plugin';

advanceTo(new Date(2020, 10, 30));

jest.setTimeout(10 * 1000);

const CONTEXT = path.join(__dirname, '../../test/package/app');

describe('webpack plugin', () => {
  test('default config', (done) => {
    expect.assertions(3);

    const compiler = webpack({
      mode: 'production',
      context: CONTEXT,
      plugins: [new BundleStatsWebpackPlugin()],
    });
    compiler.outputFileSystem = new MemoryFS();

    compiler.run((error, stats) => {
      expect(error).toEqual(null);
      expect(stats.hasErrors()).toBe(false);
      const { assets } = stats.toJson({ source: false, assets: true });
      const bundleStatsAsset = assets.find((asset) => asset.name.match(/bundle-stats\.html$/));
      expect(bundleStatsAsset).toBeTruthy();
      done();
    });
  });

  test('baseline', (done) => {
    expect.assertions(4);

    const compiler = webpack({
      mode: 'production',
      context: CONTEXT,
      plugins: [new BundleStatsWebpackPlugin({ baseline: true })],
    });
    compiler.outputFileSystem = new MemoryFS();

    compiler.run((error, stats) => {
      expect(error).toEqual(null);
      expect(stats.hasErrors()).toBe(false);
      const { assets } = stats.toJson({ source: false, assets: true });
      const bundleStatsAsset = assets.find((asset) => asset.name.match(/bundle-stats\.html$/));
      expect(bundleStatsAsset).toBeTruthy();
      const baselineAsset = assets.find((asset) =>
        asset.name.match(/node_modules\/\.cache\/bundle-stats\/baseline\.json$/),
      );
      expect(baselineAsset).toBeTruthy();
      done();
    });
  });

  describe('compare', () => {
    // Create default baseline file
    const outputDir = path.join(__dirname, '../../node_modules/.cache/bundle-stats');
    const baselineFilepath = path.join(outputDir, 'baseline.json');

    beforeEach(async () => {
      await fs.mkdir(outputDir, { recursive: true });
      await fs.writeFile(
        baselineFilepath,
        JSON.stringify({
          assets: [
            {
              name: 'main.js',
              size: 20,
            },
          ],
        }),
      );

      jest.spyOn(global.console, 'info');
      jest.spyOn(global.console, 'warn');
    });

    afterEach(async () => {
      await fs.rm(baselineFilepath);
    });

    test('should compare with baseline.json', (done) => {
      expect.assertions(6);

      const compiler = webpack({
        mode: 'production',
        context: CONTEXT,
        plugins: [new BundleStatsWebpackPlugin({ compare: true })],
      });
      compiler.outputFileSystem = new MemoryFS();

      compiler.run((error, stats) => {
        expect(error).toEqual(null);
        expect(stats.hasErrors()).toBe(false);

        /* eslint-disable no-console */
        expect(console.warn).not.toBeCalledWith('Missing baseline stats, see "baseline" option.');
        expect(console.info).toBeCalledWith(
          'Read baseline from ../node_modules/.cache/bundle-stats/baseline.json',
        );
        expect(console.info).toBeCalledWith('Bundle Size â€” 27B (+35%).');
        /* eslint-enable no-console */

        const { assets } = stats.toJson({ source: false, assets: true });
        const bundleStatsAsset = assets.find((asset) => asset.name.match(/bundle-stats\.html$/));
        expect(bundleStatsAsset).toBeTruthy();
        done();
      });
    });
  });
});
