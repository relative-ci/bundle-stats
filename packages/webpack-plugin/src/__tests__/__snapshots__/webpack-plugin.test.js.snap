// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webpack plugin baseline 1`] = `
Array [
  Object {
    "auxiliaryChunkIdHints": Array [],
    "auxiliaryChunkNames": Array [],
    "auxiliaryChunks": Array [],
    "cached": false,
    "chunkIdHints": Array [],
    "chunkNames": Array [],
    "chunks": Array [],
    "comparedForEmit": false,
    "emitted": true,
    "filteredRelated": undefined,
    "info": Object {
      "development": true,
      "size": 331477,
    },
    "isOverSizeLimit": false,
    "name": "bundle-stats.html",
    "related": Object {},
    "size": 331477,
    "type": "asset",
  },
  Object {
    "auxiliaryChunkIdHints": Array [],
    "auxiliaryChunkNames": Array [],
    "auxiliaryChunks": Array [],
    "cached": false,
    "chunkIdHints": Array [],
    "chunkNames": Array [],
    "chunks": Array [],
    "comparedForEmit": false,
    "emitted": true,
    "filteredRelated": undefined,
    "info": Object {
      "development": true,
      "size": 300,
    },
    "isOverSizeLimit": false,
    "name": "../node_modules/.cache/bundle-stats/baseline.json",
    "related": Object {},
    "size": 300,
    "type": "asset",
  },
  Object {
    "auxiliaryChunkIdHints": Array [],
    "auxiliaryChunkNames": Array [],
    "auxiliaryChunks": Array [],
    "cached": false,
    "chunkIdHints": Array [],
    "chunkNames": Array [
      "main",
    ],
    "chunks": Array [
      179,
    ],
    "comparedForEmit": false,
    "emitted": true,
    "filteredRelated": undefined,
    "info": Object {
      "javascriptModule": false,
      "minimized": true,
      "size": 27,
    },
    "isOverSizeLimit": false,
    "name": "main.js",
    "related": Object {},
    "size": 27,
    "type": "asset",
  },
]
`;

exports[`webpack plugin default config 1`] = `
Array [
  Object {
    "auxiliaryChunkIdHints": Array [],
    "auxiliaryChunkNames": Array [],
    "auxiliaryChunks": Array [],
    "cached": false,
    "chunkIdHints": Array [],
    "chunkNames": Array [],
    "chunks": Array [],
    "comparedForEmit": false,
    "emitted": true,
    "filteredRelated": undefined,
    "info": Object {
      "development": true,
      "size": 331477,
    },
    "isOverSizeLimit": false,
    "name": "bundle-stats.html",
    "related": Object {},
    "size": 331477,
    "type": "asset",
  },
  Object {
    "auxiliaryChunkIdHints": Array [],
    "auxiliaryChunkNames": Array [],
    "auxiliaryChunks": Array [],
    "cached": false,
    "chunkIdHints": Array [],
    "chunkNames": Array [
      "main",
    ],
    "chunks": Array [
      179,
    ],
    "comparedForEmit": false,
    "emitted": true,
    "filteredRelated": undefined,
    "info": Object {
      "javascriptModule": false,
      "minimized": true,
      "size": 27,
    },
    "isOverSizeLimit": false,
    "name": "main.js",
    "related": Object {},
    "size": 27,
    "type": "asset",
  },
]
`;
